### ISSUE TITLE
关于finetune_lm损失函数的问题

### ISSUE THREAD

**Issue by mali19064** [TsinghuaAI/CPM-Finetune#22]
请问在cross_entropy.py中的第38行logits.sub_(logits_max.unsqueeze(dim=-1))，为什么要将logit减去一个最大值？另外75行，定义了损失函数为loss = torch.log(sum_exp_logits) - predicted_logits，其中- predicted_logits容易理解，为标签序列的交叉熵，但我不太理解将torch.log(sum_exp_logits)加入损失函数的意义是什么? 个人理解, torch.log(sum_exp_logits)应该等价于 -log(Pmax),在损失函数中加入此项，将使得当前step输出的概率最大那个词尽可能维持较大的概率，但同时又让标签里对应的词也获得较大概率，请问我的理解是否正确？如果这样做的话，目的是为了避免灾难性遗忘还是什么？该思路有没有相关的文献出处？

**Comment by t1101675** [comment]
1. 减去最大值的原因是保证计算稳定性。因为 logits 需要过一层 softmax，里面有取 e 指数的操作，如果原来的数字过大可能会出现计算精度的问题，所以要减掉最大值。
2. `torch.log(sum_exp_logits)` 是所有 logits 取 e 指数之后再求和，再对整体取 e 指数的一项。对应着 log softmax 公式里面的第一项：
![image](https://user-images.githubusercontent.com/38183190/111060893-2a75ce80-84db-11eb-968a-f572e63de892.png)

这个 cross entropy 的实现和 torch 的实现应该是等价的，除了支持了模型并行的操作。
